SOLR-15474: Support Pluggable Circuit Breakers (#193)
apache/solr: solr/solr-ref-guide/src/circuit-breakers.adoc

It is up to the client to handle this error and potentially build a retrial logic as this should ideally be a transient situation.

== Circuit Breaker Configurations

> All circuit breaker configurations are listed as independent `<circuitBreaker>` entries in solrconfig.xml as shown below:

Reviewer: post merge conflict resolving tweak:
```suggestion
All circuit breaker configurations are listed as independent `<circuitBreaker>` entries in `solrconfig.xml` as shown below:
```

[source,xml]
----

>   <circuitBreaker class="org.apache.solr.util.circuitbreaker.CPUCircuitBreaker" enable="true">
>     <int  name="threshold">75</int>
>   </circuitBreaker>

Reviewer: minor: keep existing indentation level
```suggestion
<circuitBreaker class="org.apache.solr.util.circuitbreaker.CPUCircuitBreaker" enable="true">
  <int  name="threshold">75</int>
</circuitBreaker>
```

----

> Each circuit breaker can be independently enabled/disabled using the corresponding flag.

Reviewer 2: I'm confused about this because it looks like we don't read the flag anywhere anymore.

Reviewer: 2/2: Alternative wording might be this:

```suggestion
Each circuit breaker plugin can be independently enabled/disabled using the `enable` attribute.
```

Or to remove the sentence and the `enable="true"` mentions in the examples if/since by defaults plugins would presumably be enabled?

Another possibility would be

```
<circuitBreaker class="Foobar" enable="${circuitBreaker.foobar.enable:true}">
  ...
</circuitBreaker>
```

style examples on the (unconfirmed) assumption that this system property usage is supported for these plugins, and this would be a way for users to change the behaviour of an instance without a config change.

Author: I would disagree on the proposed XML format -- seems too clumsy to me. We could remove the enable flag altogether, although it exists for quick experimentation (toggle on/off) when using circuit breakers.

In addition, it is a good practice to exponentially back off while retrying requests on a busy node.

== Adding New Circuit Breakers
To add new circuit breakers, add the new circuit breaker's class and required parameters:

[source,xml]
----

> <circuitBreaker class="foobar" enabled="true">
>  <bool name="enabled">true</bool>
>  <int  name="threshold">75</int>
> </circuitBreaker>

Reviewer: ```suggestion
<circuitBreaker class="foobar" enabled="true">
 <int  name="threshold">75</int>
</circuitBreaker>
```

I wonder if the `enabled` flag could be removed in favour of the `enabled` attribute?
